File: python/287. Find the Duplicate Number.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        low=1
        high=len(nums)-1
        while low<high:
            mid=(low+high)/2
            count=sum(x<=mid for x in nums)
            if count>mid:
                high=mid
            else:
                low=mid+1
        return low

File: java/287. Find the Duplicate Number.java
public class Solution {
    public int findDuplicate(int[] nums) {
        return f(nums,0,nums.length-1);
    }
    public int f(int[] nums,int s,int e){
        if (s==e) {
            return 0;
        }
        if (s+1==e) {
            if (nums[s]==nums[e]) {
                return nums[s];
            }
            return 0;
        }
        int m=(s+e)/2;
        int a=f(nums,s,m);
        int b=f(nums,m+1,e);
        if (a!=0) {
            return a;
        }
        if (b!=0) {
            return b;
        }
        for (int i = s; i <=m; i++) {
            for (int j = m+1; j <=e; j++) {
                if (nums[i]==nums[j]) {
                    return nums[j];
                }
            }
        }
        return 0;
    }
}